File: oem-rnapp/ios/WeixinModule.m
//
//  WeixinModule.m
//  bankapp
//
//  Created by 陆民 on 2018/7/31.
//  Copyright © 2018年 Facebook. All rights reserved.
//

#import <Foundation/Foundation.h>

#import <React/RCTConvert.h>
#import "WeixinModule.h"

static NSString *const kOpenURLNotification = @"RCTOpenURLNotification";

@implementation WeixinModule
RCT_EXPORT_MODULE();

- (instancetype)init
{
  self = [super init];
  if (self) {
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(handleOpenURL:) name:kOpenURLNotification object:nil];
  }
  return self;
}

+ (BOOL)requiresMainQueueSetup {
  return YES;
}

- (void)handleOpenURL:(NSNotification *) notification {
  NSString *urlString = notification.userInfo[@"url"];
  NSURL *url = [NSURL URLWithString:urlString];
  [WXApi handleOpenURL:url delegate:self];
}

- (void)onResp:(BaseResp *)resp {
  if([resp isKindOfClass:[PayResp class]]){
    PayResp *payResp = (PayResp *)resp;
    
    [self sendEventWithName:@"wxpayreturn" body:@{
                                              @"returnKey": payResp.returnKey,
                                              @"type": [NSNumber numberWithInt:payResp.type],
                                              @"errCode": [NSNumber numberWithInt:payResp.errCode],
                                              @"errStr": payResp.errStr == nil ? NSNull.null : payResp.errStr
                                              }];
  }
}

- (NSArray<NSString *> *)supportedEvents {
  return @[@"wxpayreturn"];
}

RCT_EXPORT_METHOD(sendPayReq: (NSDictionary *) req)
{
  PayReq *request = [[PayReq alloc] init];
  request.partnerId = [RCTConvert NSString:req[@"partnerId"]];
  request.prepayId = [RCTConvert NSString:req[@"prepayId"]];
  request.package = @"Sign=WXPay";
  request.nonceStr = [RCTConvert NSString:req[@"nonceStr"]];
  request.timeStamp = [RCTConvert uint64_t:req[@"timeStamp"]];
  request.sign = [RCTConvert NSString:req[@"sign"]];
  
  [WXApi sendReq: request];
}

RCT_EXPORT_METHOD(sendMessageReq: (NSDictionary *) req)
{
  SendMessageToWXReq* request = nil;
  
  id mediaObject = nil;
  BOOL bText = NO;
  if(req[@"webpageUrl"] != nil) {
    WXWebpageObject* object = [WXWebpageObject object];
    object.webpageUrl = req[@"webpageUrl"];
    mediaObject = object;
    bText = NO;
  }
  
  if(mediaObject != nil) {
    request = [[SendMessageToWXReq alloc] init];
    WXMediaMessage* message = [WXMediaMessage message];
    message.mediaObject = mediaObject;
    message.title = req[@"title"];
    message.description = req[@"description"];
    if(req[@"thumbImage"] != nil) {
      UIImage* image = [UIImage imageWithData:[NSData dataWithContentsOfURL:[NSURL URLWithString:[RCTConvert NSString:req[@"thumbImage"]]]]];
      [message setThumbImage:image];
    }
    
    request.bText = bText;
    request.message = message;
    request.scene = [RCTConvert int:req[@"scene"]];
  }
  
  
  if(request != nil) {
    [WXApi sendReq: request];
  }
}

@end

File: oem-rnapp/android/app/src/main/java/com/hello/app/WeixinModule.java
package com.hello.app;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.UrlQuerySanitizer;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;
import com.tencent.mm.opensdk.modelbase.BaseResp;
import com.tencent.mm.opensdk.modelmsg.SendMessageToWX;
import com.tencent.mm.opensdk.modelmsg.WXMediaMessage;
import com.tencent.mm.opensdk.modelmsg.WXWebpageObject;
import com.tencent.mm.opensdk.modelpay.PayReq;
import com.tencent.mm.opensdk.modelpay.PayResp;
import com.tencent.mm.opensdk.openapi.IWXAPI;
import com.tencent.mm.opensdk.openapi.WXAPIFactory;

import java.io.IOException;
import java.net.URL;
import java.net.URLConnection;


public class WeixinModule extends ReactContextBaseJavaModule {

    private IWXAPI api;

    static public WeixinModule instance = null;

    public WeixinModule(ReactApplicationContext reactContext) {
        super(reactContext);
        String app_id = reactContext.getResources().getString(R.string.wx_app_id);
        api = WXAPIFactory.createWXAPI(reactContext, null);
        api.registerApp(app_id);
        instance = this;
    }

    @Override
    public String getName() {
        return "WeixinModule";
    }

    @ReactMethod
    public void sendPayReq(ReadableMap req) {
        PayReq request = new PayReq();
        request.appId = req.getString("appId");
        request.partnerId = req.getString("partnerId");
        request.prepayId = req.getString("prepayId");
        request.packageValue = "Sign=WXPay";
        request.nonceStr = req.getString("nonceStr");
        request.timeStamp = Integer.toString(req.getInt("timeStamp"));
        request.sign = req.getString("sign");
        api.sendReq(request);
    }

    @ReactMethod
    public void sendMessageReq(ReadableMap req) {
        SendMessageToWX.Req request = null;

        WXMediaMessage.IMediaObject mediaObject = null;
        if(req.getString("webpageUrl") != null) {
            WXWebpageObject object = new WXWebpageObject();
            object.webpageUrl = req.getString("webpageUrl");
            mediaObject = object;
        }

        if(mediaObject != null) {
            request = new SendMessageToWX.Req();
            WXMediaMessage message = new WXMediaMessage(mediaObject);
            message.title = req.getString("title");
            message.description = req.getString("description");
            if(req.getString("thumbImage") != null) {
                try {
                    URLConnection conn = (new URL(req.getString("thumbImage"))).openConnection();
                    conn.connect();
                    Bitmap image = BitmapFactory.decodeStream(conn.getInputStream());
                    message.setThumbImage(image);
                }catch (IOException e) {

                }

            }

            request.message = message;
            request.scene = req.getInt("scene");
        }

        if(request != null) {
            api.sendReq(request);
        }
    }

    public void onResp(BaseResp baseResp) {

        if(baseResp instanceof PayResp) {
            PayResp payResp = (PayResp) baseResp;
            WritableMap params = Arguments.createMap();

            params.putInt("errCode", payResp.errCode);
            params.putString("errStr", payResp.errStr);
            params.putString("returnKey", payResp.returnKey);

            getReactApplicationContext()
                    .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                    .emit("wxpayreturn", params);
        }



    }
}
